trade-finance-manager-ui/server/controllers/case/cancellation/reason-for-cancelling.controller.post.test.ts

```diff
@@ -0,0 +1,244 @@
+import { createMocks } from 'node-mocks-http';
+import { DEAL_SUBMISSION_TYPE } from '@ukef/dtfs2-common';
+import { aTfmSessionUser } from '../../../../test-helpers';
+import { postReasonForCancelling, PostReasonForCancellingRequest } from './reason-for-cancelling.controller';
+import { PRIMARY_NAVIGATION_KEYS } from '../../../constants';
+import { ReasonForCancellingErrorsViewModel, ReasonForCancellingViewModel } from '../../../types/view-models';
+import api from '../../../api';
+import { canSubmissionTypeBeCancelled } from '../../helpers';
+
+const validateReasonForCancellingMock: jest.Mock<ReasonForCancellingErrorsViewModel> = jest.fn(() => ({
+  errorSummary: [],
+  reasonForCancellingErrorMessage: undefined,
+}));
+
+jest.mock('./validation/validate-reason-for-cancelling', () => ({
+  validateReasonForCancelling: (reason: string | undefined) => validateReasonForCancellingMock(reason),
+}));
+
+jest.mock('../../../api', () => ({
+  getDeal: jest.fn(),
+  updateDealCancellation: jest.fn(),
+}));
+
+jest.mock('../../helpers', () => ({
+  canSubmissionTypeBeCancelled: jest.fn(() => true),
```

Comment:
Can we test without mocking this helper? I.e pass different mock data into the helper rather than mocking the helper's response

Reply:
I think as a unit test, it makes sense to mock other dependencies

Reply:
The `canSubmissionTypeBeCancelled` helper is tested in its own unit test

Reply:
I would recommend testing the function without mocking for an effecting and a better coverage of the helper function.

Reply:
I can't see any reason to mock this helper in the unit test (or any helper in a unit test really that is not async) - can we please update it so it's not mocked? E.g

```ts

// in test 1 - deal can be cancelled
const mockXDeal = {};
jest.mocked(api.getDeal).mockResolvedValue(mockXDeal);


// in test 2 - deal cannot be cancelled
const mockYDeal = {};
jest.mocked(api.getDeal).mockResolvedValue(mockYDeal);
```

Then we don't need things like this:
```ts
jest.mock('../../helpers', () => ({
  canSubmissionTypeBeCancelled: jest.fn(() => true),
}));

jest.mocked(canSubmissionTypeBeCancelled).mockReturnValueOnce(false);
```

Reply:
Same with `validateReasonForCancelling` as well @MarRobSoftwire , I can't see any reason to mock these.

The controllers consume/call these functions, the tests can just call these with different parms/arguments, instead of mocking the other functions responses. I would say that by _not_ doing this, it's not representative of what the controller is actually doing

Reply:
Can we please replace the mocks for the following @BethThomas141 

```ts
jest.mock('./validation/validate-reason-for-cancelling'
jest.mock('../../helpers'
```

Same principle as comment from @abhi-markan  here https://github.com/UK-Export-Finance/dtfs2/pull/3571#discussion_r1780772262

Reply:
@ttbarnes I've now replaced the mock for `canSubmissionTypeBeCancelled` 👍 

I'd be keen to keep mocking `validateReasonForCancelling` - the reason for this one, is it's a more complicated function call and the unit tests for our controller here should just check that the validation response we get back is displayed and formatted correctly on the Nunjucks page. The actual message that we get is tested in the unit test for `validate-reason-for-cancelling.test.ts` so doesn't need re-testing here. 

If we don't mock it here, every time we change any of the logic in `validate-reason-for-cancelling.ts` we will have to change the unit tests in _both_ `validate-reason-for-cancelling.test.ts` and the tests here, touching two files instead of one. 

Would you be ok with a compromise of removing the mock for  `canSubmissionTypeBeCancelled`, but keeping the mock in for `validateReasonForCancelling`? 🙂

Reply:
Thanks @BethThomas141 I see the point - the issue I find with the mocking of the validation (`validateReasonForCancelling`) (or any other type functions called in a unit test, excluding async/await calls), is that essentially:

What happens outside of unit tests:
1. Controller receives a body.
2. Controller calls the validation function with that body.
3. Controller does XYZ if there are or aren't validation errors.

But the test is doing this:

1. Controller receives a body.
2. Validation function returns whatever we mock, instead of it's true logic, based on what is in the payload.
  a. This can also lead to misleading mocks/tests.
  b. Currently you can pass in an invalid payload and it'll still pass - it'd expect it to only pass with a valid payload.
4. Controller does XYZ if there are or aren't validation errors.

We could have an e.g `validBody` object and 1x simple mock object with errors or even just an empty object and then remove the mock,. Then it's mimicking what the controller is actually doing - calling the validation function instead of mocking/stubbing it.

There are lots of mock instances I think we should remove/replace, but one for another day. Therefore happy to go with current approach, resolving

Reply:
Thanks Tony - appreciate you writing that out. I can see pros and cons of both sides here, it's a tricky one. Thanks for resolving, happy to revisit the discussion in future.